plugins/bnr-dp/fu-bnr-dp-device.c

```diff
@@ -0,0 +1,746 @@
+	active_header = fu_bnr_dp_device_fw_header(self, FU_BNR_DP_MODULE_NUMBER_RECEIVER, error);
+	if (active_header == NULL)
+		return NULL;
+	if (!fu_bnr_dp_firmware_patch_boot_counter(
+		FU_BNR_DP_FIRMWARE(firmware),
+		fu_struct_bnr_dp_payload_header_get_counter(active_header),
+		error))
+		return NULL;
+
+	/* check fw image */
+	factory_data = fu_bnr_dp_device_factory_data(self, FU_BNR_DP_MODULE_NUMBER_RECEIVER, error);
+	if (factory_data == NULL)
+		return NULL;
+	fw_header = fu_struct_bnr_dp_payload_header_parse(g_bytes_get_data(bytes, NULL),
+							  g_bytes_get_size(bytes),
+							  FU_BNR_DP_FIRMWARE_HEADER_OFFSET,
+							  error);
+	if (fw_header == NULL)
+		return NULL;
+	if (!fu_bnr_dp_firmware_check(FU_BNR_DP_FIRMWARE(firmware),
+				      factory_data,
+				      active_header,
+				      fw_header,
+				      error))
+		return NULL;
+
+	return g_steal_pointer(&firmware);
+}
+
+static gboolean
+fu_bnr_dp_device_write_firmware(FuDevice *device,
+				FuFirmware *firmware,
+				FuProgress *progress,
+				FwupdInstallFlags flags,
+				GError **error)
+{
+	FuBnrDpDevice *self = FU_BNR_DP_DEVICE(device);
+	g_autoptr(GBytes) bytes = NULL;
+	g_autoptr(GByteArray) read_back = NULL;
+
+	/* progress, values based on dev tests with -vv */
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_add_flag(progress, FU_PROGRESS_FLAG_GUESSED);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_WRITE, 34, "write");
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_READ, 65, "verify");
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_RESTART, 1, "activate");
+
+	/* get payload bytes including patched boot counter */
+	bytes = fu_firmware_get_bytes_with_patches(firmware, error);
+	if (bytes == NULL)
+		return FALSE;
+
+	/* write new firmware to inactive area */
+	if (!fu_bnr_dp_device_write_data(self,
+					 FU_BNR_DP_OPCODES_FLASH_USER,
+					 FU_BNR_DP_MODULE_NUMBER_RECEIVER,
+					 0,
+					 g_bytes_get_data(bytes, NULL),
+					 g_bytes_get_size(bytes),
+					 fu_progress_get_child(progress),
+					 error))
+		return FALSE;
+	fu_progress_step_done(progress);
+
+	/* verify written data */
+	read_back = fu_bnr_dp_device_read_data(self,
+					       FU_BNR_DP_OPCODES_FLASH_USER,
+					       FU_BNR_DP_MODULE_NUMBER_RECEIVER,
+					       0,
+					       FU_BNR_DP_FIRMWARE_SIZE,
+					       fu_progress_get_child(progress),
+					       error);
+	if (read_back == NULL)
+		return FALSE;
+	if (!fu_memcmp_safe(g_bytes_get_data(bytes, NULL),
+			    g_bytes_get_size(bytes),
+			    0,
+			    read_back->data,
+			    read_back->len,
+			    0,
+			    FU_BNR_DP_FIRMWARE_SIZE,
+			    error)) {
+		g_prefix_error_literal(error, "verification of written firmware failed: ");
+		return FALSE;
+	}
+	fu_progress_step_done(progress);
+
+	/* apply new firmware by resetting the device */
+	if (!fu_bnr_dp_device_reset(self, FU_BNR_DP_MODULE_NUMBER_RECEIVER, error))
+		return FALSE;
+	/* give controller some time before ->reload() tries to read info again */
```

Comment:
does the dpaux node "go away and come back"? or is it just unresponsive for this time?

Reply:
> that hot plug doesn't work

We can certainly *use* those to rescan -- e.g. check out the synaptics_mst plugin.

Reply:
I can reproduce with synaptics_mst -- debugging.